chainstate/src/detail/mod.rs

```diff
@@ -271,52 +274,179 @@ impl<S: BlockchainStorage, V: TransactionVerificationStrategy> Chainstate<S, V>
+    {
         let mut attempt_number = 0;
         loop {
-            log::info!("Processing block: {}", block_id);
+            log::info!("Processing block ({}), block id = {}", comment, block_id);
 
             let mut chainstate_ref = self.make_db_tx().map_err(BlockError::from).log_err()?;
 
-            let best_block_id = chainstate_ref
-                .get_best_block_id()
-                .map_err(BlockError::BestBlockLoadError)
-                .log_err()?;
-
-            chainstate_ref
-                .check_block(&block, false)
-                .map_err(BlockError::CheckBlockFailed)
-                .log_err()?;
+            let result = func(&mut chainstate_ref).log_err()?;
 
-            let block_index = chainstate_ref.persist_block(&block).log_err()?;
-            let result =
-                chainstate_ref.activate_best_chain(block_index, best_block_id).log_err()?;
             let db_commit_result = chainstate_ref.commit_db_tx().log_err();
             match db_commit_result {
-                Ok(_) => {}
+                Ok(_) => return Ok(result),
                 Err(err) => {
                     if attempt_number >= *self.chainstate_config.max_db_commit_attempts {
                         return Err(BlockError::DatabaseCommitError(
-                            block.get_id(),
+                            *block_id,
                             *self.chainstate_config.max_db_commit_attempts,
                             err,
+                            comment.into(),
                         ));
                     } else {
                         attempt_number += 1;
                         continue;
                     }
                 }
             }
+        }
+    }
+
+    /// This is similar to `with_rw_tx_for_block_id`, but it also maintains a mutable state that
+    /// is passed to `func` and then returned to the caller.
+    /// Note that the state is reset to `initial_state` on each commit attempt and the returned
+    /// state is the one from the last commit attempt.
+    fn with_rw_tx_and_state_for_block_id<Func, Res, State>(
+        &mut self,
+        mut func: Func,
+        initial_state: &State,
+        block_id: &Id<Block>,
+        comment: &str,
+    ) -> (State, Result<Res, BlockError>)
+    where
+        Func: FnMut(
+            &mut chainstateref::ChainstateRef<TxRw<'_, S>, V>,
+            &mut State,
+        ) -> Result<Res, BlockError>,
+        State: Clone,
+    {
+        let mut state = initial_state.clone();
+
+        let result = self.with_rw_tx_for_block_id(
+            |chainstate_ref| func(chainstate_ref, &mut state),
+            block_id,
+            comment,
+        );
+
+        (state, result)
+    }
 
-            return Ok(result);
+    // Perform all the regular block checks, updating `block_status` after each successful check.
+    // The returned bool indicates whether a reorg has occurred.
+    fn perform_block_checks(
```

Comment:
I wouldn't call this just "checks"... it's more. I don't know whether we should just call this "put_in_blocktree" or something similar to signal that this "mutates" the state, unlike just checking.

Reply:
I've renamed it to `integrate_block`, let me know, if this is ok.

Reply:
OK for now. Thank you.